add/selenium-tests: selenium tests that replicate the existing e2e tests

BoardJames opened the pull request:
## Description
Since Cypress only supports Chrome (and derivatives) currently and it has some other limitations to do with tab navigation and selection I decided to explore Selenium testing. 

It still has limitations - notably actions like Shift+Click or key combinations like Ctrl+a don't seem to be possible on Firefox because Selenium web-driver for node does not support the new W3C web driver specification yet.

Run in Chrome:
```bash
npm run test-browser
```

Run in Firefox:
```bash
test_browser=firefox npm run test-browser
```

Note in theory you should be able to change the test_browser environment variable to specify other browsers like ie, edge and safari but I have not tried it yet.

## How Has This Been Tested?
The tests have been run on Firefox and Chrome on Linux Mint.

## Screenshots (jpeg or gifs if applicable):
![selenium-tests-demo](https://user-images.githubusercontent.com/26243021/33255276-848b81b4-d398-11e7-8bd6-ab50ae39848c.gif)


## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

tiny-james committed:
add/selenium-tests: selenium tests that replicate the existing e2e tests

youknowriad commented:
I'm aware Cypress has limitations but I personally don't think we should use several frameworks for our tests. If we were to change, I think we should change all our tests and from previous experiences, Selenium adds a lot of burden and maintainability issues because of random errors. I think we should just hold on testing those areas for now or try to test the smaller bits using unit tests.

Not saying we shouldn't look for alternatives for Cypress in the future if we find it doesn't do the job properly but It has proven to be stable and pretty easy to maintain and work with IMO.

EphoxJames commented:
> I'm aware Cypress has limitations but I personally don't think we should use several frameworks for our tests. If we were to change, I think we should change all our tests and from previous experiences, Selenium adds a lot of burden and maintainability issues because of random errors.  I think we should just hold on testing those areas for now or try to test the smaller bits using unit tests.

I have already ported all the existing Cypress tests so we can switch to Selenium completely. It was actually a pretty nice experience. The main point of trickiness was loading up the new page because I had to dismiss a confirmation popup and ensure that the editor was set to visual mode. Things are more verbose but I enjoy the control that it gives you. The only errors I have received so far have not been "random" but were due to changes in page structure (ie changed class names) and they would have also needed to be changed in Cypress.

> Not saying we shouldn't look for alternatives for Cypress in the future if we find it doesn't do the job properly but It has proven to be stable and pretty easy to maintain and work with IMO.

We're already hit major limitations in Cypress. @ephox-mogran tried to test keyboard navigation with it and found it could not send a Tab key ( https://github.com/cypress-io/cypress/issues/299 - it looks like this may have been fixed now). I tried to test selection with it and found that typing does not set the selection realistically. I can not test the behavior of Autocomplete with it after spending a day trying and it took me 20minutes to get the test working using Selenium. Weirdly Cypress does not actually seem to be testing the behaviour of Chrome - it is testing the behavior of Cypress which just happens to run in Chrome... Cypress is rapidly evolving at the moment. It may turn out to be a great environment for testing in future but it is insufficient at the moment.

Another counterpoint is that Cypress only tests Chrome (and derivatives). The majority of the developers already test in Chrome every day because that is the primary browser used for development so we really need automated tests that can run on other browsers like Edge, Firefox and Safari.

tiny-james committed:
Merge branch 'master' into add/selenium-tests

tiny-james committed:
add/selenium-tests: update css selectors to match master

tiny-james committed:
add/selenium-tests: Add test to check that all blocks can be created

tiny-james committed:
Merge branch 'master' into add/selenium-tests

tiny-james committed:
add/selenium-tests: Increased login timeout to account for slow server start

tiny-james committed:
add/selenium-tests: Restructure blockTitles calculation to be readable

tiny-james committed:
Merge branch 'master' into add/selenium-tests

tiny-james committed:
Merge branch 'master' into add/selenium-tests

tiny-james committed:
add/selenium-tests: Small test fixes

- Modified quote test to account for new <cite> tag
- Match "Code Editor" & "Visual Editor" menu items with regex to ignore
the keyboard shortcut
- Add a delay to login so the onload javascript does not change focus
while typing

tiny-james committed:
add/selenium-tests: Added test of using inline patterns

tiny-james committed:
add/selenium-tests: Bumped version numbers of dependencies to latest

tiny-james committed:
Merge branch 'master' into add/selenium-tests

tiny-james committed:
add/selenium-tests: ported 004-managing-links from e2e tests

tiny-james committed:
add/selenium-tests: Renamed added tests to increment numbering

tiny-james committed:
add/selenium-tests: Tweaked inline-patterns test to handle browser differences

tiny-james committed:
Merge branch 'master' into add/selenium-tests

youknowriad commented:
Valid points about cypress here, we're going with Google puppeteer right now. Let's give it try and see how it goes.

Thanks for your time trying alternatives.

#5618
